### PR TITLE
Build and upload deb packages for Debian and Ubuntu

### PR DESCRIPTION
**Pull request by nekronos**
This PR builds deb packages for all supported Debian and Ubuntu versions. When a version tag is pushed, in the [semantic versioning](https://semver.org/) format, a release will be created and the packages uploaded as release assets. An example of how this will look can be viewed [here](https://github.com/Pinata-Consulting/OpenROAD/releases).

A deb package provides a convenient way to distribute and install OpenROAD. And to make this feature useful you should consider introducing a process for making regular releases. I am ready to assist if desired.

Consider this PR a starting point/suggestion for introducing regular releases and packages.

----

Depends on https://github.com/The-OpenROAD-Project/OpenROAD/pull/2705

### CODE REVIEW

**Comment by oharboe** [comment]
Nice. Does it include the Yosys binaries as well?

The Yosys binaries for OpenROAD would need to live side by side by official Yosys packages.

Perhaps by prefixing yosys? 

or-yosys and or-yosys-abc.

Are there any other binaries needed for OpenROAD-flow-scripts?

**Comment by QuantamHD** [comment]
Klayout is also required, but already supported in debian

**Comment by nekronos** [comment]
> Nice. Does it include the Yosys binaries as well?
> 
> The Yosys binaries for OpenROAD would need to live side by side by official Yosys packages.
> 
> Perhaps by prefixing yosys?
> 
> or-yosys and or-yosys-abc.
> 
> Are there any other binaries needed for OpenROAD-flow-scripts?

This does not include yosys. Only the binaries built by cmake in this repository are included. The dependencies must be defined so that we can make the appropriate packaging (bundling with the openroad deb or separate deb etc)

**Comment by oharboe** [comment]
> > Nice. Does it include the Yosys binaries as well?
> > The Yosys binaries for OpenROAD would need to live side by side by official Yosys packages.
> > Perhaps by prefixing yosys?
> > or-yosys and or-yosys-abc.
> > Are there any other binaries needed for OpenROAD-flow-scripts?
> 
> This does not include yosys. Only the binaries built by cmake in this repository are included. The dependencies must be defined so that we can make the appropriate packaging (bundling with the openroad deb or separate deb etc)

I don't know if yosys and berkley abc are patched by OpenROAD, or if it is just the latest version of yosys. Certainly, Ubuntu 22.04 has an older version of Yosys than OpenROAD uses.

**Comment by maliberty** [comment]
There is an open task to update yosys in ORFS.  It just needs some work to qualify the new results are good.  In general yosys advances very little for asic synthesis so updating rarely helps much.

We have no generally been doing releases and advise people to use head.  This could change if it is a way to get automated releases.  Is tagging in git the only requirement?  What is the latency between a tag and when users will see the build?

**Comment by nekronos** [comment]
> There is an open task to update yosys in ORFS. It just needs some work to qualify the new results are good. In general yosys advances very little for asic synthesis so updating rarely helps much.
> 
> We have no generally been doing releases and advise people to use head. This could change if it is a way to get automated releases. Is tagging in git the only requirement? What is the latency between a tag and when users will see the build?

The github actions I have included in this PR will create a release whenever a tag is manually pushed to github. The tag will trigger a build of the `.deb` packages and with the speed of github actions buildes it takes around ~100 minutes for the packages to be built and uploaded to github.

Also, the `debian/changelog` file should be updated on every release. This can be automated.

The build times can be reduced significantly by not rebuilding the docker images on every release and using selfhosted/faster builders for building the `.deb` packages.

Tagging could also be handled by automation. For example if nightly/weekly builds are useful, a cron job can be set up to push tags.

**Comment by maliberty** [comment]
Please change the trigger to be manual.  Distributing GPL binaries is an issue for us.  We'd like to accept the PR but not have it run in this repo.

**Comment by nekronos** [comment]
> Please change the trigger to be manual. Distributing GPL binaries is an issue for us. We'd like to accept the PR but not have it run in this repo.

@vvbandeira I have changed it to be manually triggered only. And when triggered the `.deb` packages will be uploaded as github actions workflow artifacts instead of creating a release.

File: .github/workflows/github-actions-build-deb-package.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  #push:
+  #  tags:
+  #    - v*
+
+  #pull_request:
+  #  paths:
+  #    - 'debian/changelog'
+  #    - 'etc/DependencyInstaller.sh'
+  #    - 'docker/*'
+  #    - '.github/workflows/github-actions-build-deb-package.yml'
```
[the remark this answers is not in the document]
Comment (nekronos) [reply .github/workflows/github-actions-build-deb-package.yml]: Removing the tag one, is will not make a release anymore. But lets keep the pull_request trigger as it shows what files should trigger a build

**Review by vvbandeira — APPROVED** [review]

**Pull request merged**